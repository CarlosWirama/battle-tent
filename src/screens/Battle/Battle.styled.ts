import styled from 'styled-components';
import { Centered } from '../../components/basics';

export const LayoutContainer = styled(Centered)`
  background-image: linear-gradient(141deg,#1294a2 35%,#0a4861 75%);
  justify-content: flex-start;
  height: 100vh;
  color: white;
`;

export const TopArea = styled.header`
  display: flex;
  justify-self: flex-start;
  justify-content: space-between;
  width: 100%;
  box-sizing: border-box;
  padding: 16px;
`;

export const OpponentAvatar = styled.div`
  display: inline-block;
  height: 32px;
  width: 32px;
  border-radius: 50%;
  border: 2px solid white;
  padding: 16px;
`;
export const OpponentInfo = styled.div`
  display: flex;
  justify-content: flex-start;
`;
