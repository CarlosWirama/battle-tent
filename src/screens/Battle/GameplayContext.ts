import React, { Dispatch, SetStateAction } from 'react';
import Pokemon from '../../types/Pokemon';

export default React.createContext<{
  party: Pokemon[];
  setParty: Dispatch<SetStateAction<Pokemon[]>>
}>({
  party: [],
  setParty: () => { },
});
