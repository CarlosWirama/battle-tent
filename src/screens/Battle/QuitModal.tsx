import React, { useContext } from 'react';
import { Button } from '../../components/basics';
import Modal, { ModalProps } from '../../components/Modal';
import { NavigationContext, ScreenState } from '../../navigation';

export function QuitModal({ shown, onClose }: Omit<ModalProps, 'children'>) {
  const navigate = useContext(NavigationContext);

  function quit() {
    navigate(ScreenState.Lobby);
  }

  return (
    <Modal shown={shown} onClose={onClose}>
      Do you really wish to quit?
      <Button onClick={quit}>Yes</Button>
      <Button onClick={onClose}>No</Button>
    </Modal>
  );
}
