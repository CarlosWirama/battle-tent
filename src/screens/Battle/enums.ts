export enum BattleStep {
  ChooseParty,
  ChooseMoves,
}
