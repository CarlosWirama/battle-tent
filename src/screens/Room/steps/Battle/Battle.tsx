import React, { useState, useContext, useEffect, useRef } from 'react';
import GamplayContext from '../../GameplayContext';
import {
  emitEndTurn,
  subscribeTurnChanged,
  subscribeMoveUsed,
  emitUseMove,
} from '../../../../api';
import { Button } from '../../../../components/basics';
import { Move } from '../../../../types/Pokemon';
import { PlayerContext } from '../../../../auth';
import {
  BattleArea,
  PartyArea,
  PartyTile,
  TileDetail,
  HealthBar,
  MoveOptionBox,
  MoveTile,
  EnergyBarContainer,
  EnergyBar,
} from './Battle.styled';
import {
  animateAttacking,
  animateTakingDamage,
} from './animate';
import { getPokemonModel } from '../../../../components/PokemonModel/helper';
import { concat } from 'ramda';

type NullableIdx = number | null;

export default function Battle() {
  const [energy, setEnergy] = useState(0);
  const [maxEnergy, setMaxEnergy] = useState(1);
  const [availableMoves, setAvailableMoves] = useState<Move[][]>([]);
  const [choosenMoveIdx, setChoosenMoveIdx] = useState<NullableIdx>(null);
  const [choosenPokemonIdx, setChoosenPokemonIdx] = useState<NullableIdx>(null);
  const [choosenOpponentIdx, setChoosenOpponentIdx] = useState<NullableIdx>(null);

  const partyTileRef = [
    useRef<HTMLImageElement>(null),
    useRef<HTMLImageElement>(null),
    useRef<HTMLImageElement>(null),
  ];
  const opponentTileRef = [
    useRef<HTMLImageElement>(null),
    useRef<HTMLImageElement>(null),
    useRef<HTMLImageElement>(null),
  ];

  const { party, opponentParty, updateParties } = useContext(GamplayContext);
  const [player] = useContext(PlayerContext);

  // subscription
  useEffect(function subscription() {

    const sTurnChanged = subscribeTurnChanged(({ my_turn, energy, moves }) => {
      setEnergy(my_turn ? energy : 0);
      setMaxEnergy(my_turn ? energy : 1);
      setAvailableMoves(my_turn ? moves : []);
    });

    const sMoveUsed = subscribeMoveUsed(({
      move,
      userIndex,
      moveIndex,
      remainingEnergy,
      parties,
      targetIndexes: { myParty, opponentParty } = {},
    }) => {
      // animate user and targets
      console.log(`${userIndex} used ${move.name}!`);
      if (myParty && opponentParty) {
        console.log(`${myParty.concat(opponentParty)} affected`);
      }
      console.log(`remainingEnergy: ${remainingEnergy}`);
      setEnergy(remainingEnergy);
      updateParties(parties);
    });

    return function unsubscribe() {
      sTurnChanged.off();
      sMoveUsed.off();
    }
  }, [player, updateParties]);

  function animateMove(move: Move, userIndex: number, targetIndexes?: number[]) {
    const partyTileElement = partyTileRef[userIndex].current;
    animateAttacking(partyTileElement);
    targetIndexes?.forEach(index => {
      const opponentTileElement = opponentTileRef[index].current;
      animateTakingDamage(opponentTileElement);
    });
  }

  function onClickOpponentPokemon(index: number) {
    if (choosenPokemonIdx === null || choosenMoveIdx === null) {
      return;
    }
    animateMove(
      availableMoves[choosenMoveIdx][choosenMoveIdx],
      choosenPokemonIdx,
      [index]
    );
    setEnergy(energy => energy - 1);
    const moveTargetMultipleOpponent = false;
    let targetIndexes = [index];
    if (moveTargetMultipleOpponent) {
      if (choosenOpponentIdx === null) {
        setChoosenOpponentIdx(index);
        return;
      }
      targetIndexes = [choosenOpponentIdx, index];
    }
    emitUseMove({
      userMoveIndex: [choosenPokemonIdx, choosenMoveIdx],
      targetIndexes: targetIndexes,
    });
    // TODO: make move unusable
    setChoosenOpponentIdx(null);
    setChoosenMoveIdx(null);
  }

  function endTurn() {
    emitEndTurn();
  }

  const energyBar = concat(
    new Array(energy).fill({ empty: false }),
    new Array(maxEnergy - energy).fill({ empty: true })
  );
  return (
    <>
      <BattleArea>
        <PartyArea style={{ alignSelf: 'flex-end' }}>
          {opponentParty.map(({ health, maxHealth, pokemon: { name } }, index) => (
            <PartyTile
              chosen={choosenOpponentIdx === index}
              onClick={() => onClickOpponentPokemon(index)}
              key={index}
            >
              <img
                src={getPokemonModel(name)}
                alt={name}
                ref={opponentTileRef[index]}
              />
              <TileDetail>
                <div>{name}</div>
                <HealthBar percentage={health / maxHealth * 100} />
              </TileDetail>
            </PartyTile>
          ))}
        </PartyArea>
        <PartyArea>
          {party.map(({ health, maxHealth, pokemon: { name } }, index) => (
            <PartyTile
              chosen={choosenPokemonIdx === index}
              onClick={() => setChoosenPokemonIdx(index)}
              key={index}
            >
              <img
                src={getPokemonModel(name, 'back')}
                alt={name}
                ref={partyTileRef[index]}
              />
              <TileDetail>
                <div>{name}</div>
                <HealthBar percentage={health / maxHealth * 100} />
              </TileDetail>
            </PartyTile>
          ))}
        </PartyArea>
      </BattleArea>
      <MoveOptionBox>
        {choosenPokemonIdx === null || availableMoves[choosenPokemonIdx] === undefined
          ? null
          : availableMoves[choosenPokemonIdx].map((move, index) => {
            const isDisabled = move.energy > energy;
            return (
              <MoveTile
                chosen={choosenMoveIdx === index}
                onClick={() => isDisabled || setChoosenMoveIdx(index)}
                key={index}
                disabled={isDisabled}
              >
                <div>{move.name}</div>
                <div>Power {move.power} - PP {move.pp}</div>
              </MoveTile>
            );
          })
        }
      </MoveOptionBox>
      <EnergyBarContainer>
        {energyBar.map(({ empty }, index) => (
          <EnergyBar empty={empty} key={index} />
        ))}
      </EnergyBarContainer>
      <Button onClick={endTurn}>End Turn</Button>
    </>
  );
}

export type ChosenItem = null | {
  itemId: string;
  partyIndex?: number;
};
