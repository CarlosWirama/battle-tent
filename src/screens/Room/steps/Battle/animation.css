@keyframes taking-damage {
  0% { opacity: 1 }
  10% { opacity: 0 }
  50% { opacity: 1 }
  60% { opacity: 0 }
}

.taking-damage {
  animation: taking-damage .1s 1 alternate;
}

@keyframes attacking {
  0% { transform: none }
  100% { transform: translate(16px, -16px) rotate(12deg); }
}

@keyframes attacking-down {
  0% { transform: none }
  100% { transform: translate(-16px, 16px) rotate(-12deg); }
}

.attacking {
  animation: attacking .1s 1 alternate;
}

.attacking.attack-direction-down {
  animation: attacking-down .1s 1 alternate;
}

@keyframes returning-to-pokeball {
  20% {
    filter:
      invert(100%)
      sepia(100%)
      saturate(0%)
      hue-rotate(184deg)
      brightness(1000%)
      contrast(103%);
    transform: none;
  }
  30% { transform: scale(1, 1); }
  75% { transform: scale(.3, .3); }
  98% { transform: scale(.3, .3); }
  100% {
    filter:
      invert(100%)
      sepia(100%)
      saturate(0%)
      hue-rotate(184deg)
      brightness(1000%)
      contrast(103%);
    transform: scale(0, 0);
  }
}

.returning-to-pokeball {
  animation: returning-to-pokeball 1s 1 forwards;
}

.fainted {
  visibility: hidden;
}
