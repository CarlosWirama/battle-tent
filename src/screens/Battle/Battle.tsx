import React, { useState, useEffect } from 'react';
import { Button } from '../../components/basics';
import LoadingIndicator from '../../components/LoadingIndicator';
import * as Steps from './steps';
import { QuitModal } from './QuitModal';
import { fetchOpponent, fetchPokemonList } from '../../apis/battleApi';
import { BattleStep } from './enums';
import Pokemon from '../../types/Pokemon';
import Trainer from '../../types/Trainer';
import { LayoutContainer, TopArea, OpponentInfo, OpponentAvatar } from './Battle.styled';

export default function Battle() {
  const [isLoading, setIsLoading] = useState(false);
  const [quitModalShown, setQuitModalShown] = useState(false);
  const [activeStep, setActiveStep] = useState(BattleStep.ChooseParty);
  const [pokemonList, setPokemonList] = useState<Pokemon[]>([]);
  const [opponent, setOpponent] = useState<Trainer | undefined>(undefined);

  useEffect(() => {
    setIsLoading(true);
    Promise.all([
      getOpponent(),
      getPokemonList(),
    ]).finally(() => setIsLoading(false));
  }, []);

  async function getOpponent() {
    try {
      const response = await fetchOpponent();
      setOpponent(response.data);
    } catch (e) {
      console.error(e);
      alert(e);
    }
  }

  async function getPokemonList() {
    try {
      const response = await fetchPokemonList();
      setPokemonList(response.data);
    } catch (e) {
      console.error(e);
      alert(e);
    }
  }

  function openQuitModal() {
    setQuitModalShown(true);
  }
  function closeQuitModal() {
    setQuitModalShown(false);
  }

  function Step() {
    function ChooseParty() {
      return (
        <Steps.ChooseParty
          pokemonList={pokemonList}
          setActiveStep={setActiveStep}
        />
      );
    }
    switch (activeStep) {
      case BattleStep.ChooseParty: return <ChooseParty />;
      case BattleStep.ChooseMoves: return <Steps.ChooseMoves />;
      default: return <ChooseParty />;
    }
  }

  return (
    <LayoutContainer>
      <QuitModal shown={quitModalShown} onClose={closeQuitModal} />
      <TopArea>
        <OpponentInfo>
          <OpponentAvatar />
          <div>{opponent && opponent.name}</div>
        </OpponentInfo>
        <Button onClick={openQuitModal}>Quit</Button>
      </TopArea>
      <main>
        {isLoading ? <LoadingIndicator /> : <Step />}
      </main>
    </LayoutContainer>
  );
}
