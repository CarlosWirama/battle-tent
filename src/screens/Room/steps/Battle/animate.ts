import './animation.css';

// enum Animation { Idle, TakingDamage, Attacking }

function refreshElementAnimation(
  animationClass: string,
  element: HTMLDivElement
) {
  element.classList.remove(animationClass);
  void element.offsetWidth;
}

function animateOnce(animationClass: string, element: HTMLDivElement) {
  element.classList.add(animationClass);
  setTimeout(() => refreshElementAnimation(animationClass, element), 200);
}

export function animateTakingDamage(element: HTMLDivElement | null) {
  if (element === null) return;
  animateOnce('taking-damage', element);
}

export function animateAttacking(element: HTMLDivElement | null) {
  if (element === null) return;
  animateOnce('attacking', element);
}
