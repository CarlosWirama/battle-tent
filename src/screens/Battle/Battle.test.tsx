import React from 'react';
import { render } from '@testing-library/react';
import Battle from './Battle';

describe('Battle screen', () => {
  it('should render', () => {
    const { asFragment } = render(<Battle />);
    expect(asFragment()).toMatchSnapshot();
  });
});
