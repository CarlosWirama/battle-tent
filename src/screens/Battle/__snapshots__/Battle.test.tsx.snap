// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Battle screen should render 1`] = `
<DocumentFragment>
  <div
    class="sc-AxirZ sc-fzozJi fzcSCU"
  >
    <header
      class="sc-fzoLsD wHbcZ"
    >
      <div
        class="sc-fzplWN hcOeEo"
      >
        <div
          class="sc-fzpans huHwMF"
        />
        <div />
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text sc-AxjAm faKiNC"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Quit
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </header>
    <main>
      <div
        class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
